packages/@aws-cdk/aws-ec2/lib/cfn-init-elements.ts

```diff
@@ -0,0 +1,405 @@
+  }
+}
+
+/**
+ * Context information passed when an InitElement is being consumed
+ */
+export interface InitBindOptions {
+  /**
+   * Scope in which to define any resources, if necessary.
+   */
+  readonly scope: Construct;
+
+  /**
+   * Which OS platform (Linux, Windows) the init block will be for.
+   * Impacts which config types are available and how they are created.
+   */
+  readonly platform: InitPlatform;
+
+  /**
+   * Ordered index of current element type. Primarily used to auto-generate
+   * command keys and retain ordering.
+   */
+  readonly index: number;
+
+  /**
+   * Instance role of the consuming instance or fleet
+   */
+  readonly instanceRole: iam.IRole;
+}
+
+/**
+ * A return type for a configured InitElement. Both its CloudFormation representation, and any
+ * additional metadata needed to create the CloudFormation:Init.
+ */
+export interface InitElementConfig {
+  /**
+   * The CloudFormation representation of the configuration of an InitElement.
+   */
+  readonly config: Record<string, any>;
+
+  /**
+   * Optional authentication blocks to be associated with the Init Config
+   *
+   * @default - No authentication associated with the config
+   */
+  readonly authentication?: Record<string, any>;
+}
+
+/**
+ * Base class for all CloudFormation Init elements
+ */
+export abstract class InitElement {
+
+  /**
+   * Returns the init element type for this element.
+   */
+  public abstract readonly elementType: InitElementType;
+
+  /**
+   * Called when the Init config is being consumed. Renders the CloudFormation
+   * representation of this init element, and calculates any authentication
+   * properties needed, if any.
+   *
+   * @param options bind options for the element.
+   */
+  public abstract bind(options: InitBindOptions): InitElementConfig;
+
+}
+
+/**
+ * Options for InitCommand
+ */
+export interface InitCommandOptions {
+  /**
+   * Identifier key for this command
+   *
+   * You can use this to order commands.
+   *
+   * @default - Automatically generated
+   */
+  readonly key?: string;
+
+  /**
+   * Sets environment variables for the command.
+   *
+   * This property overwrites, rather than appends, the existing environment.
+   *
+   * @default - Use current environment
+   */
+  readonly env?: Record<string, string>;
+
+  /**
+   * The working directory
+   *
+   * @default - Use default working directory
+   */
+  readonly cwd?: string;
+
+  /**
+   * Command to determine whether this command should be run
+   *
+   * If the test passes (exits with error code of 0), the command is run.
+   *
+   * @default - Always run this command
+   */
+  readonly test?: string;
+
+  /**
+   * Continue running if this command fails
+   *
+   * @default false
+   */
+  readonly ignoreErrors?: boolean;
+
+  /**
+   * Specifies how long to wait (in seconds) after a command has finished in case the command causes a reboot.
+   *
+   * A value of "forever" directs cfn-init to exit and resume only after the
```

Comment:
Doesn't look like the user can specify 'forever' here.

Reply:
Doesn't feel right.

We usually support all the options that CloudFormation provides, so if 'forever' is an option that the user can specify in CF, we should support that.

There are several ways to do this. One way that comes to mind (names illustrative) -

```
readonly waitAfterCompletion?: WaitDuration;

export class WaitDuration {
  public static of(duration: Duration): WaitDuration;

  public static forever(): WaitDuration;
}
```

Or with two properties here and synth time checks that they are consistent (less favourable).

There may be other ones as well.

Reply:
I'd prefer a `waitForever?: boolean` flag that is mutex with the current field than an entirely different type. Yes, the extra type is "more correct" but it's also MORE HASSLE for the common case.